jgt_common/__init__.py

```diff
@@ -872,6 +872,86 @@ def check_until(
     raise IncompleteAtTimeoutException(msg, call_result=result, timeout=timeout)
 
 
+@classify("misc", "exceptions")
+def assert_if_values(format_if_format, error_fun=lambda x: "\n".join(truths_from(x))):
+    """
+    Assert if any truthy values are returned (or yielded) from the decorated function.
+
+    Any values that the decorated function yields/returns are passed to ``error_fun``.
+    ``format_if_format``, and the result of the ``error_fun`` call,
+    are passed to ``format_if``.
+    The result of the ``format_if`` call is used for an assertion.
+
+    Args:
+        format_if_format (str): the first parameter to format_if.
+        error_fun (callable): called on the results of the decorated function,
+            return value will be passed as the 2nd parameter to format_if.
+            The default function will only process truthy values.
+            That can be changed by passing in your own custom ``error_fun``
+
+    Returns:
+        None - If no values are returned, or yielded,
+        or the result of ``error_fun`` is falsey, return None.
+
+    Raises:
+        AssertionError: as described above.
+
+    This decorator is meant to simplify code that otherwise has to manually keep
+    track of a list of things to complain about::
+
+        errors = []
+        for thing in things:
+           if bad(thing):
+               errors.append(description_of(thing))
+        err_msg = format_if("blah blah blah: {}", "".join(errors))
+        assert not err_msg, err_msg
+
+    into simpler code::
+
+        @assert_if_values("blah blah blah: {}")
+        ...
+            for thing in things:
+                if bad(thing):
+                    yield description_of(thing)
+
+    """
+
+    @_wrapt.decorator
+    def helper(wrapped, instance, args, kwargs):
+        err_msg = format_if(format_if_format, error_fun(wrapped(*args, **kwargs)))
+        assert not err_msg, err_msg
+
+    return helper
+
+
+# NOTE: Not using wrapt. as wrapt would add another level of nesting / complexity..
+#       The intent is dynamic run-time wrapping, not static decorating of, a function.
+#       If that changes, then this could change to use wrapt later without changing
+#       the users of this function.
+
+
+@classify("misc", "sequence")
+def accumulator_for(fun):
+    """
+    Accumulate the results of calling fun into a list, and return that list.
+
+    Note: If you use this is a straight-up decorator,
+    then all invocations of the function will accumulate into one list.
+    The intention is that you use this as an on-the-spot wrapper,
```

Comment:
NB: I love this function, but the docstring took me a few minutes of head-scratching to figure out what we're doing here. Maybe we could pare down the `check_until` reference note and squeeze in an example?

Reply:
Hmmm, yeah, definitely a case of "Well, _I_ knew what I was talking about!" :-)
I did think about using `check_until` in an actual example, but `check_until` while super helpful for what it does do, isn't easy to use for a pithy example...
Will keep thinking on this.